Translates Suspense for Data Fetching (Experimental) (#358)
reactjs/pt-BR.reactjs.org: content/docs/concurrent-mode-suspense.md

---
id: concurrent-mode-suspense
title: Suspense para Busca de Dados (Experimental)
permalink: docs/concurrent-mode-suspense.html
prev: concurrent-mode-intro.html
next: concurrent-mode-patterns.html
---

>Cuidado:
>
>Esta página descreve **recursos experimentais que [ainda não estão disponíveis](/docs/concurrent-mode-adoption.html) em uma versão estável**. Não confie nas versões experimentais do React em aplicativos de produção. Esses recursos podem mudar significativamente e sem aviso antes de se tornarem parte do React.
>
>Esta documentação é destinada a adotantes precoces e pessoas curiosas. Se você é novo no React, não se preocupe com esses recursos -- você não precisa aprendê-los agora.

O React 16.6 adicionou um componente `<Suspense>` isso permite que você "esperar" para que algum código seja carregado e especifique declarativamente um state de carregamento (como um spinner) enquanto esperamos:

```jsx

> const ProfilePage = React.lazy(() => import('./ProfilePage')); // Carregado de forma preguiçosa

Author: Traduzi como **Carregado quando necessário**, acredito que dá o sentido correto, o que você acha?

- [O Que É Suspense, Exatamente?](#what-is-suspense-exactly)
  - [O Que Suspense Não É](#what-suspense-is-not)
  - [O Que O Suspense Permite Que Você Faça](#what-suspense-lets-you-do)
- [Usando Suspense na Prática](#using-suspense-in-practice)
  - [E Se Eu Não Uso Relay?](#what-if-i-dont-use-relay)
  - [Para Autores de Biblioteca](#for-library-authors)
- [Abordagens Tradicionais vs Suspense](#traditional-approaches-vs-suspense)
  - [Abordagem 1: Busca-na-Renderização (sem usar Suspense)](#approach-1-fetch-on-render-not-using-suspense)
  - [Abordagem 2: Busca-Depois-da-Renderização (sem usar Suspense)](#approach-2-fetch-then-render-not-using-suspense)
  - [Abordagem 3: Renderização-Conforme-Você-Busca (usando Suspense)](#approach-3-render-as-you-fetch-using-suspense)
  - [Ainda Estamos Descobrindo Isso](#were-still-figuring-this-out)
- [Suspense e Race Conditions](#suspense-and-race-conditions)

>   - [Race Conditions com useEffect](#race-conditions-with-useeffect)

Author: Acho que poderia ser sim **condição de corrida**, mas não acha ficaria melhor **condição de concorrência**? Por passar o sentido real dessa expressão?

}
```

Vamos comparar como diferentes estratégias de busca de dados lidam com esse requisito.

### Race Conditions com `useEffect` {#race-conditions-with-useeffect}

> Primeiro, tentaremos uma versão do nosso exemplo original de "busca em efeito" example. Vamos modificá-lo para passar um parâmetro `id` das props de `<ProfilePage>` para `fetchUser(id)` e `fetchPosts(id)`:

Reviewer: ```suggestion
Primeiro, tentaremos uma versão do nosso exemplo original de "busca no efeito". Vamos modificá-lo para passar um parâmetro `id` das props de `<ProfilePage>` para `fetchUser(id)` e `fetchPosts(id)`:
```

}
```

**[Experimente no CodeSandbox](https://codesandbox.io/s/nervous-glade-b5sel)**

> Observe como também mudamos as dependências do efeito de `[]` para `[id]` — porque queremos que o efeito seja executado novamente quando o `id` mudar. Caso contrário, não buscaremos novamente novos dados.

Reviewer: ```suggestion
Observe como também mudamos as dependências do efeito de `[]` para `[id]` — porque queremos que o efeito seja executado novamente quando o `id` mudar. Caso contrário, não buscaremos novamente os dados.
```

> Se tentarmos esse código, pode parecer que ele funcione primeiro. No entanto, se aleatorizarmos o tempo de atraso em nossa implementação "fake API" e pressionarmos o botão "Next" com rapidez suficiente, veremos nos logs do console que algo está saindo muito errado. **Às vezes, as requisições dos perfis anteriores podem "voltar" depois que já mudamos o perfil para outro ID -- e, nesse caso, elas podem sobrescrever o novo state com uma resposta obsoleta para um ID diferente.**

Reviewer: ```suggestion
Se tentarmos esse código, a princípio pode parecer que ele funcione. No entanto, se aleatorizarmos o tempo de atraso em nossa implementação da "fake API" e pressionarmos o botão "Next" com rapidez suficiente, veremos nos logs do console que algo está saindo muito errado. **Às vezes, as requisições dos perfis anteriores podem "voltar" depois que já mudamos o perfil para outro ID -- e, nesse caso, elas podem sobrescrever o novo state com uma resposta obsoleta para um ID diferente.**
```

}
```

**[Experimente no CodeSandbox](https://codesandbox.io/s/trusting-clarke-8twuq)**

Este código é enganosamente fácil de ler.

Infelizmente, nem o uso de uma classe nem a sintaxe `async` / `await` nos ajudaram a resolver esse problema. Esta versão sofre exatamente as mesmas race conditions, pelas mesmas razões.

### O Problema {#the-problem}

> Os componentes do React possuem seu próprio "ciclo de vida". Eles podem receber props atualizar o state a qualquer momento. No entanto, cada requisição assíncrona *também* possui seu próprio "ciclo de vida". Começa quando iniciamos e termina quando obtemos uma resposta. A dificuldade que estamos enfrentando é "sincronizar" vários processos no tempo que se afetam. Isso é difícil de pensar.

Reviewer: ```suggestion
Os componentes do React possuem seu próprio "ciclo de vida". Eles podem receber props ou atualizar o state a qualquer momento. No entanto, cada requisição assíncrona *também* possui seu próprio "ciclo de vida". Começa quando iniciamos e termina quando obtemos uma resposta. A dificuldade que estamos enfrentando é "sincronizar" vários processos no tempo que se afetam. Isso é difícil de pensar.
```

}
```

**[Experimente no CodeSandbox](https://codesandbox.io/s/infallible-feather-xjtbu)**

> No exemplo anterior utilizando Suspense, tínhamos apenas um `recurso`, portanto mantivemos em uma variável de nível superior. Agora que temos vários recursos, o movemos para o state do componente `<App>`:

Reviewer: Acredito que `resource` não deve ser traduzido, já que é um recurso do código acima, o que acha? Também vi em outros lugares ele ter sido traduzido, se for corrigir isso dê uma pesquisada no restante da página.

}
```

**[Experimente no CodeSandbox](https://codesandbox.io/s/adoring-goodall-8wbn7)**

> Ele pegaria erros de renderização *e* erros da busca de dados do Suspense. Podemos ter o error boundaries que quisermos, mas é melhor [ser intencional](https://aweary.dev/fault-tolerance-react/) sobre o posicionamento deles.

Reviewer: ```suggestion
Ele pegaria erros de renderização *e* erros da busca de dados do Suspense. Podemos ter quantos error boundaries quisermos, mas é melhor [ser intencional](https://aweary.dev/fault-tolerance-react/) sobre o posicionamento deles.
```

## Próximos Passos {#next-steps}

Agora, abordamos o básico do Suspense para Busca de Dados! Importante, agora entendemos melhor *por que* o Suspense funciona dessa maneira e como ele se encaixa no espaço de busca de dados.

> O Suspense responde a algumas perguntas, mas também coloca novas questões por conta própria:

Reviewer: ```suggestion
O Suspense responde a algumas perguntas, mas também trás suas próprias novas questões:
```
